File: tools/py/commit_aggregator.py
"""
Commit 聚合工具 - 批量聚合和Token优化

功能说明：
- 同类 commit 聚合
- Token 优化
- 智能过滤

使用方法：
    python tools/py/commit_aggregator.py [--max-count N] [--since DATE]

参数说明：
    --max-count N           最多聚合N个commit（默认：50）
    --since DATE            聚合指定日期之后的commit
    --optimize-token        启用Token优化

版本信息：
    版本：1.0.0
    更新日期：2025-12-11
"""

import json
import argparse
import time
from typing import List, Dict


def aggregate_by_type(commits: List[Dict]) -> Dict:
    """按类型聚合commits"""
    by_type = {}
    
    for commit in commits:
        commit_type = commit.get('type', 'unknown')
        if commit_type not in by_type:
            by_type[commit_type] = []
        by_type[commit_type].append(commit)
    
    return by_type


def optimize_tokens(commits: List[Dict]) -> Dict:
    """优化Token消耗"""
    # 统计原始Token
    original_tokens = sum(
        len(c.get('what', '')) + len(c.get('why', '')) + 
        sum(len(h) for h in c.get('how', []))
        for c in commits
    )
    
    # 聚合同类型
    by_type = aggregate_by_type(commits)
    
    # 生成摘要
    summaries = []
    for commit_type, type_commits in by_type.items():
        summary = {
            'type': commit_type,
            'count': len(type_commits),
            'changes': [c.get('what', '') for c in type_commits[:5]]  # 只保留前5个
        }
        summaries.append(summary)
    
    # 统计优化后Token
    optimized_tokens = sum(
        len(s['type']) + sum(len(c) for c in s['changes'])
        for s in summaries
    )
    
    reduction = ((original_tokens - optimized_tokens) / original_tokens * 100) if original_tokens > 0 else 0
    
    return {
        'summaries': summaries,
        'original_tokens': original_tokens,
        'optimized_tokens': optimized_tokens,
        'reduction_percent': round(reduction, 2)
    }


def main():
    start_time = time.time()
    
    parser = argparse.ArgumentParser(description='Commit Aggregator')
    parser.add_argument('--max-count', type=int, default=50)
    parser.add_argument('--since', help='Since date')
    parser.add_argument('--optimize-token', action='store_true')
    args = parser.parse_args()
    
    # 获取commits
    from commit_parser import get_commits
    commits = get_commits(args.max_count, args.since)
    
    if args.optimize_token:
        result_data = optimize_tokens(commits)
    else:
        result_data = aggregate_by_type(commits)
    
    elapsed_time = round(time.time() - start_time, 4)
    result = {
        'data': result_data,
        'metadata': {
            'elapsed_seconds': elapsed_time,
            'version': '1.0.0'
        }
    }
    
    print(json.dumps(result, indent=2, ensure_ascii=False))


if __name__ == '__main__':
    main()

File: tools/js/commit_aggregator.js
/**
 * Commit 聚合工具 - 批量聚合和Token优化
 * 
 * 功能说明：
 * - 同类 commit 聚合
 * - Token 优化
 * 
 * 使用方法：
 *   node tools/js/commit_aggregator.js [--max-count N] [--optimize-token]
 * 
 * 参数说明：
 *   --max-count N           最多聚合N个commit（默认：50）
 *   --optimize-token        启用Token优化
 * 
 * 版本信息：
 *   版本：1.0.0
 *   更新日期：2025-12-11
 */

const { execSync } = require('child_process');

function getCommits(maxCount, since = null) {
    try {
        let cmd = `git log --max-count=${maxCount} --pretty=format:"%H|%an|%ad|%s" --date=iso`;
        if (since) {
            cmd += ` --since="${since}"`;
        }
        
        const output = execSync(cmd, { encoding: 'utf8' });
        
        return output.split('\n')
            .filter(line => line)
            .map(line => {
                const [hash, author, date, subject] = line.split('|');
                // 模拟简单的类型解析，实际应调用 commit_parser 或匹配其逻辑
                let type = 'other';
                if (subject.toLowerCase().startsWith('feat')) type = 'feature';
                else if (subject.toLowerCase().startsWith('fix')) type = 'fix';
                else if (subject.toLowerCase().startsWith('docs')) type = 'docs';
                else if (subject.toLowerCase().startsWith('refactor')) type = 'refactor';
                
                return { 
                    hash, 
                    author,
                    date,
                    what: subject, 
                    type: type,
                    why: '', // 简化版，不解析 body
                    how: []  // 简化版
                };
            });
    } catch {
        return [];
    }
}

function optimizeTokens(commits) {
    const originalTokens = commits.reduce((sum, c) => sum + (c.what || '').length + (c.why || '').length, 0);
    
    const byType = {};
    commits.forEach(c => {
        const type = c.type || 'unknown';
        if (!byType[type]) byType[type] = [];
        byType[type].push(c);
    });
    
    const summaries = Object.entries(byType).map(([type, typeCommits]) => ({
        type,
        count: typeCommits.length,
        changes: typeCommits.slice(0, 5).map(c => c.what)
    }));
    
    const optimizedTokens = summaries.reduce(
        (sum, s) => sum + s.type.length + s.changes.reduce((s2, c) => s2 + c.length, 0), 
        0
    );
    
    const reduction = originalTokens > 0 
        ? ((originalTokens - optimizedTokens) / originalTokens * 100).toFixed(2)
        : 0;
    
    return {
        summaries,
        original_tokens: originalTokens,
        optimized_tokens: optimizedTokens,
        reduction_percent: parseFloat(reduction)
    };
}

function main() {
    const startTime = Date.now();
    
    const args = process.argv.slice(2);
    let maxCount = 50;
    let since = null;
    let optimizeToken = false;
    
    for (let i = 0; i < args.length; i++) {
        if (args[i] === '--max-count') maxCount = parseInt(args[++i]);
        else if (args[i] === '--since') since = args[++i];
        else if (args[i] === '--optimize-token') optimizeToken = true;
    }
    
    const commits = getCommits(maxCount, since);
    const resultData = optimizeToken ? optimizeTokens(commits) : { commits };
    
    const elapsedTime = ((Date.now() - startTime) / 1000).toFixed(4);
    const result = {
        data: resultData,
        metadata: {
            elapsed_seconds: parseFloat(elapsedTime),
            version: '1.0.0'
        }
    };
    
    console.log(JSON.stringify(result, null, 2));
}

main();
